hca/dss/__init__.py

```diff
@@ -107,87 +109,182 @@ def download(self, bundle_uuid, replica, version="", dest_name="",
+                                content = response.iter_content(chunk_size=bytes_to_read)
+                                chunk = next(content)
+                                if chunk:
+                                    consume_bytes -= len(chunk)
+
+                        for chunk in response.iter_content(chunk_size=1024*1024):
+                            if chunk:
+                                fh.write(chunk)
+                                hasher.update(chunk)
+                                retries_left = min(retries_left + 1, num_retries)
+                                delay = max(delay / 2, min_delay_seconds)
+                        break
+                    finally:
+                        response.close()
+                except (ChunkedEncodingError, ConnectionError, ReadTimeout):
+                    if retries_left > 0:
+                        logger.info("%s", "File {}: GET FAILED. Attempting to resume.".format(
+                            file_uuid, dest_path))
+                        time.sleep(delay)
+                        delay *= 2
+                        retries_left -= 1
+                        continue
+                    raise
+
+        if hasher.hexdigest().lower() != file_sha256.lower():
+            os.remove(dest_path)
+            logger.error("%s", "File {}: GET FAILED. Checksum mismatch.".format(file_uuid))
+            raise ValueError("Expected sha256 {} Received sha256 {}".format(
+                file_sha256.lower(), hasher.hexdigest().lower()))
+        else:
+            logger.info("%s", "File {}: GET SUCCEEDED. Stored at {}.".format(file_uuid, dest_path))
+
+    @classmethod
+    def _file_path(cls, checksum):
+        """
+        returns a file's relative local path based on the nesting parameters and the files hash
+        :param checksum: a string checksum
+        :return: relative Path object
+        """
+        assert(sum(cls.FILE_NESTING_PARAMETER) <= len(checksum))
+        checksum_index = 0
+        path_pieces = []
+        for prefix_length in cls.FILE_NESTING_PARAMETER:
+            path_pieces.append(checksum[checksum_index:(checksum_index + prefix_length)])
+            checksum_index += prefix_length
+        path_pieces.append(checksum)
+        return os.path.join(*path_pieces)
+
+    @classmethod
+    def _parse_manifest(cls, manifest):
+        with open(manifest) as f:
+            # unicode_literals is on so all strings are unicode. CSV wants a str so we need to jump through a hoop.
+            delimiter = '\t'.encode('ascii') if USING_PYTHON2 else '\t'
+            reader = csv.DictReader(f, delimiter=delimiter, quoting=csv.QUOTE_NONE)
+            return list(reader)
+
+    def download_manifest_v2(self, manifest, replica, num_retries=10, min_delay_seconds=0.25):
+        """
+
+        Process the given manifest file in TSV (tab-separated values) format and download the files referenced by it.
+        The files are downloaded in the version 2 format.
```

Comment:
What is the "version 2 format" and can you include a link to it?

Reply:
Ah, that sounds reasonable.  Just confusing to me seeing this new.

Reply:
@hannes-ucsc I need some clarification on this plan before I write a description.

Reply:
Will wait for this then.  Otherwise looks good to approve.